cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/PartitionedDataLayer.java

```diff
@@ -279,12 +280,52 @@ public SSTablesSupplier sstables(int partitionId,
         LOGGER.info("Creating partitioned SSTablesSupplier for Spark partition partitionId={} rangeLower={} rangeUpper={} numReplicas={}",
                     partitionId, range.lowerEndpoint(), range.upperEndpoint(), replicas.size());
 
+        return constructSSTablesSupplier(partitionId, replicationFactor, instRanges, replicas, range);
+    }
+
+    @NotNull
+    private SSTablesSupplier constructSSTablesSupplier(
+    int partitionId,
+    ReplicationFactor replicationFactor,
+    Map<Range<BigInteger>, List<CassandraInstance>> instRanges,
+    Set<CassandraInstance> replicas,
+    Range<BigInteger> range)
+    {
+        if (consistencyLevel == ConsistencyLevel.EACH_QUORUM)
+        {
+            Map<String, Integer> minReplicasPerDC = consistencyLevel.eachQuorumBlockFor(replicationFactor);
+            LOGGER.debug("partitionId={}, minReplicasPerDC={}", partitionId, minReplicasPerDC);
+            Map<String, Set<CassandraInstance>> replicasByDC = replicas.stream()
+                                                                       .collect(Collectors.groupingBy(
+                                                                       CassandraInstance::dataCenter,
+                                                                       Collectors.toSet()
+                                                                       ));
+
+            Map<String, SSTablesSupplier> perDCSSTablesSupplier = new HashMap<>();
+            for (Map.Entry<String, Integer> entry : minReplicasPerDC.entrySet())
+            {
+                Set<CassandraInstance> replicasInDC = replicasByDC.getOrDefault(entry.getKey(), new HashSet<>());
```

Comment:
Seems like you can use empty set here to avoid instantiating each time?

```suggestion
                Set<CassandraInstance> replicasInDC = replicasByDC.getOrDefault(entry.getKey(), Collections.emptySet());
```